File: Python/DoublePoint/trap/trap.py
def tarp(heights: list[int]) -> int:
    length = len(heights)

    sum = 0
    left = 0
    right = length - 1

    leftMax = 0
    rightMax = 0

    while left < right:
        if heights[left] < heights[right]:
            leftMax = max(leftMax, heights[left])
            sum += leftMax - heights[left]
            left += 1
        else:
            rightMax = max(rightMax, heights[right])
            sum += rightMax - heights[right]
            right -= 1

    return sum

File: golang/DoublePoint/trap/trap.go
// Package trap jieyushui
package trap

func tarp(height []int) int {
	sum := 0
	length := len(height)

	left := 0
	right := length - 1

	leftMax := 0
	rightMax := 0

	for left < right {
		if height[left] < height[right] {
			leftMax = max(leftMax, height[left])
			sum += leftMax - height[left]
			left++
		} else {
			rightMax = max(rightMax, height[right])
			sum += rightMax - height[right]
			right--
		}
	}

	return sum
}
